solr/core/src/java/org/apache/solr/handler/component/CompoundQueryComponent.java

```diff
@@ -0,0 +1,146 @@
+        for (var rb_i : crb.responseBuilders) {
+          nextStage = Math.min(nextStage, super.distributedProcess(rb_i));
+        }
+      } else if (rb.getStage() == CompoundResponseBuilder.STAGE_FUSION) {
+        nextStage = doFusion(crb);
+      }
+    }
+    return nextStage;
+  }
+
+  private int doFusion(CompoundResponseBuilder crb) {
+    final SolrDocumentList responseDocs = new SolrDocumentList();
+    final TopDocs[] hits = new TopDocs[crb.responseBuilders.size()];
+    for (int crb_idx = 0; crb_idx < crb.responseBuilders.size(); ++crb_idx) {
+
+      final SolrDocumentList sdl = crb.responseBuilders.get(crb_idx).getResponseDocs();
+
+      final ScoreDoc[] scoreDocs = new ScoreDoc[sdl.size()];
+      for (int idx = 0; idx < sdl.size(); ++idx) {
+        scoreDocs[idx] = new ScoreDoc(idx /* doc */, 0f /* score */, crb_idx /* shardIndex */);
+      }
+
+      final TotalHits totalHits =
+          new TotalHits(
+              sdl.getNumFound(),
+              sdl.getNumFoundExact()
+                  ? TotalHits.Relation.EQUAL_TO
+                  : TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO);
+
+      hits[crb_idx] = new TopDocs(totalHits, scoreDocs);
+    }
+
+    final var params = crb.req.getParams();
+    final int topN = params.getInt(CommonParams.ROWS, CommonParams.ROWS_DEFAULT);
+    final int k = params.getInt("rrf.k", 1);
+
+    final TopDocs fusion = TopDocs.rrf(topN, k, hits);
+
+    for (ScoreDoc scoreDoc : fusion.scoreDocs) {
+      responseDocs.add(
+          crb.responseBuilders.get(scoreDoc.shardIndex).getResponseDocs().get(scoreDoc.doc));
+    }
+    final TotalHits totalHits = fusion.totalHits;
+    responseDocs.setNumFound(totalHits.value());
+    responseDocs.setNumFoundExact(TotalHits.Relation.EQUAL_TO.equals(totalHits.relation()));
+
+    crb.setResponseDocs(responseDocs);
+
+    return ResponseBuilder.STAGE_DONE;
+  }
+
+  @Override
+  public void handleResponses(ResponseBuilder rb, ShardRequest sreq) {
+    if (rb instanceof CompoundResponseBuilder crb) {
+      for (var rb_i : crb.responseBuilders) {
+        if (rb_i.isThisFromMe(sreq)) {
+          super.handleResponses(rb_i, sreq);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void finishStage(ResponseBuilder rb) {
+    if (rb instanceof CompoundResponseBuilder crb) {
+      for (var rb_i : crb.responseBuilders) {
+        super.finishStage(rb_i);
+      }
+
+      if (rb.getStage() == CompoundResponseBuilder.STAGE_GET_FIELDS) {
+        rb.resultIds = new HashMap<>();
+        for (var rb_i : crb.responseBuilders) {
+          for (var entry : rb_i.resultIds.entrySet()) {
+            /*
+            this positionInResponse is an interim value that will change during the fusion stage
+            but we need it to be unique here to support any SolrPluginUtils.copyNamedListIntoArrayByDocPosInResponse
+            call in HighlightComponent (an alternative would be to make the HighlightComponent aware of the fusion stage)
+            */
```

Comment:
Just proof-of-concept here, practically I think the `HighlightComponent` would need to be aware of the fusion stage, not just because of the position-in-response logic but also because the responses for the two sub-queries can overlap and then there are two (possibly different) highlights to choose from, or two highlights to merge.